# Iris Classification

- **Supervised Learning(지도학습)**  
학습을 위한 다양한 Features와 분류 결정값인 Label 데이터로 Model을 학습한 뒤, 별도의 Test Dataset에서 미지의 Label을 예측한다.  
즉, 명확한 정답이 주어진 데이터를 먼저 학습한 뒤 미지의 정답을 예측하는 방식이다.  
(예) Classification(분류)  


    - Train Dataset : 학습을 위해 주어진 데이터 세트  
    - Test Dataset : 머신러닝 모델의 예측 성능을 평가하기 위해 별도로 주어진 데이터 세트  

- scikit-learn의 다양한 모듈  
    - **sklearn.datasets** : 사이킷런에서 자체적으로 제공하는 데이터 세트를 생성하는 모듈의 모임  
    - **sklearn.model_selection** : train dataset, test dataset, prediction dataset으로 데이터를 분리(train_test_split() 함수)하거나 최적의 Hyperparameter로 평가하기 위한 다양한 모듈의 모임  
        \#Hyperparameter : 머신러닝 알고리즘별로 최적의 학습을 위해 직접 입력하는 파라미터들을 통칭함, Hyperparameter를 통해 머신러닝 알고리즘의 성능을 튜닝할 수 있음

## 붓꽃 품종 예측하기  
ML 알고리즘 : Decision Tree(의사 결정 트리) 알고리즘을 구현한 DecisionTreeClassifier를 적용

**Process**  
1. Dataset 분리 : train, test dataset을 분리한다.  
2. Model training(모델 학습) : train data를 기반으로 ML Algorithm을 적용해 모델을 학습시킨다.  
3. Make Predictions(예측 수행) : 학습된 ML 모델을 이용해 test data의 분류(붓꽃 품종)를 예측한다.  
4. Model Evaluation(모델 평가) : 예측된 결괏값과 test data의 실제 결괏값을 비교해 ML 모델 성능을 평가한다.

- Iris 데이터의 Features  
    - Sepal length (꽃받침의 길이)
    - Sepal width (꽃받침의 너비)
    - Petal length (꽃잎의 길이)
    - Petal width (꽃잎의 너비)  


- Iris 데이터의 Label(품종)
    - Setosa
    - Vesicolor
    - Virginica

### Iris Dataset 준비

In [1]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [13]:
import pandas as pd

#Iris Dataset load
iris = load_iris()

#Iris Dataset 정보 확인하기
#print(iris['DESCR'])

In [11]:
#iris.data는 Iris 데이터에서 feature만으로 된 데이터를 numpy로 가지고 있음
iris_data = iris.data
print('iris feature명: ', iris.feature_names)
print('iris feature값: ', iris_data[:5]) #5개 행만 표시

iris feature명:  ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
iris feature값:  [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]]


In [12]:
#iris.target은 Iris Dataset에서 Label(결정 값) 데이터를 numpy로 가지고 있음
iris_label = iris.target
print('iris target명: ', iris.target_names)
print('iris target값: ', iris_label)

iris target명:  ['setosa' 'versicolor' 'virginica']
iris target값:  [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]


Label(결정값)은 0, 1, 2 세 가지 값으로 돼 있으며 0이 Setosa, 1이 Versicolor, 2가 Virginica 품종을 의미한다.

In [14]:
#DataFrame으로 변환하기
iris_df = pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label'] = iris.target
iris_df.head(5)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   label  
0      0  
1      0  
2      0  
3      0  
4      0

### train, test dataset split

- train_test_split() : 학습 데이터와 테스트 데이터를 test_size 파라미터 입력 값의 비율로 쉽게 분할한다.  

In [23]:
#train_test_split() 호출하기
X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size=0.2, random_state=11)

- 입력 파라미터
1. iris_data : Feature dataset  
2. iris_label : Label dataset  
3. test_size=0.2 : 전체 dataset 중 test dataset의 비율을 20%로 함  
4. random_state : random값을 만드는 seed와 같은 의미, 수행할 때마다 동일한 dataset으로 분리하기 위해 random_state를 일정한 숫자 값으로 부여함  


- 반환 값
1. 학습용 Feature dataset을 X_train으로 반환함  
2. 테스트용 Feature dataset을 X_test로 반환함  
3. 학습용 Label dataset을 y_train으로 반환함  
4. 테스트용 Label dataset을 y_test로 반환함

### Model Training(모델 학습)

ML Classification Algorithms 중 하나인 Decision Tree(의사 결정 트리)를 이용해 학습

- fit( ) : 학습 수행 메서드  

In [24]:
#DecisionTreeClassifier 객체 생성  
dt_clf = DecisionTreeClassifier(random_state=11)

#학습 수행
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=11)

### Predictions(예측 수행)

- predict( ) : 예측 수행 메서드

In [26]:
#학습이 완료된 DecisionTreeClassifier 객체에서 테스트 데이터 세트로 예측 수행  
pred = dt_clf.predict(X_test)

### Evaluation(모델 성능 평가)

- accuracy_score( ) : 정확도 측정 메서드  
\#Accuracy(정확도) : 예측 결과가 실제 레이블 값과 얼마나 정확하게 맞는지를 평가하는 지표  
    - 입력 파라미터  
        1. 실제 Label dataset  
        2. 예측 Label dataset

In [28]:
#예측한 붓꽃 품종과 실제 테스트 데이터 세트의 붓꽃 품종이 얼마나 일치하는지 확인하기
from sklearn.metrics import accuracy_score
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도: 0.9333


학습한 의사 결정 트리의 알고리즘 예측 정확도가 약 0.9333(93.33%)으로 측정됨